src/schematools/validation.py

```diff
@@ -79,192 +57,166 @@ def __init__(self, message: str) -> None:  # noqa: D107
         self.message = message
 
 
-class Validator:
-    """Base class for validators.
+_all: List[tuple[str, Callable[[DatasetSchema], Iterator[str]]]] = []
 
-    Not only is this a base class for validators,
-    it is also used for running all registered validators.
-    See Also: :meth:`run_all`
 
-    Registration is a side-effect of overriding this base class.
+def run_all(dataset: DatasetSchema) -> Iterator[ValidationError]:
+    r"""Run all registered validators.
 
-    """
-
-    _registry: ClassVar[List[Type[Validator]]] = []
-    dataset: DatasetSchema
+    Yields:
+        :class:`ValidationError`\s, if any.
 
-    @classmethod
-    def __init_subclass__(cls: Type[Validator]) -> None:
-        """Register sub classes."""
-        super().__init_subclass__()
-        cls._registry.append(cls)
+    """  # noqa: W605
+    for name, validator in _all:
+        for msg in validator(dataset):
+            yield ValidationError(validator_name=name, message=msg)
 
-    def __init__(self, dataset: DatasetSchema) -> None:
-        """Initialize the validator with a dataset.
 
-        Args:
-            dataset: The dataset to run the validations on.
-        """
-        self.dataset = dataset
+def _validator(name: str) -> Callable:
```

Comment:
What's the rationale behind the starting `_`?

Reply:
Privacy. I'm not sure if this is considered a public module at all, but there are currently no validators outside this file. We can always remove the underscore if we want to support validators from user code (but why would we?).

Reply:
Yeah, makes sense.